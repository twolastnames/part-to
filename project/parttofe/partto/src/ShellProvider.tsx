import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "./App.scss";

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

export enum Theme {
  Default = "default",
  Dark = "dark",
}

export function changeTheme(theme: Theme) {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeChangeRef?.current && themeChangeRef.current(theme);
  themeNode.className = `${theme}Theme`;
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(Theme.Default);
  themeChangeRef = useRef(setTheme);
  return (
    <div id="theme" className="defaultTheme">
      <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
    </div>
  );
};
