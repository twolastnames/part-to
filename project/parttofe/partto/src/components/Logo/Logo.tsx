import React from "react";

import classes from "./Logo.module.scss";

export enum Size {
  Small = 60,
  Medium = 120,
  Large = 240,
}

export interface LogoProps {
  size?: Size;
}

export function Logo({ size: rawSize }: LogoProps) {
  const size = rawSize || Size.Medium;
  return (
    <div
      className={classes.logo}
      data-testid="Logo"
      style={{
        display: "inline-block",
        height: `${size * 0.5}px`,
        overflow: "hidden",
        width: `${size * 1.5}px`,
        fontFamily: '"Courier New", Courier, monospace',
        fontSize: `${size}px`,
        lineHeight: `${size * 1.5}px`,
      }}
    >
      <div
        style={{
          position: "relative",
          top: `-${size * 0.4}px`,
        }}
      >
        <div
          style={{
            height: `${size * 0.65}px`,
            overflow: "hidden",
          }}
        >
          <div>T</div>
        </div>
        <div
          style={{
            height: `${size * 0.6}px`,
            overflow: "hidden",
            position: "relative",
            top: `-${size * 0.4}px`,
            left: `${size * 0.2}px`,
          }}
        >
          <div
            style={{
              position: "relative",
              top: `-${size * 0.2}px`,
            }}
          >
            o
          </div>
        </div>
      </div>
    </div>
  );
}
