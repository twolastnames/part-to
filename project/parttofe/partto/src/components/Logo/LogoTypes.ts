export enum Size {
  Small = 60,
  Medium = 120,
  Large = 240,
}

export interface LogoProps {
  size?: Size;
}
