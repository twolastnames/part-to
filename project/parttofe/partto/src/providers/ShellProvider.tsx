import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "../App.scss";
import {
  LeftDynamicItemSetPairProvider,
  RightDynamicItemSetPairProvider,
} from "./DynamicItemSetPair";
import { TimerProvider } from "./Timer";

export enum Theme {
  Default = "default",
  Dark = "dark",
}

const themeStorageKey = "theme";

export function getTheme(): Theme {
  return (localStorage.getItem(themeStorageKey) || Theme.Default) as Theme;
}

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

function setThemeClass() {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeNode.className = `${getTheme()}Theme`;
}

export function changeTheme(theme: Theme) {
  localStorage.setItem(themeStorageKey, theme.toString());
  themeChangeRef?.current && themeChangeRef.current(theme);
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(getTheme());
  themeChangeRef = useRef(setTheme);
  setTimeout(() => {
    setThemeClass();
  }, 0);
  return (
    <div id="theme" className="defaultTheme">
      <LeftDynamicItemSetPairProvider>
        <RightDynamicItemSetPairProvider>
          <TimerProvider>
            <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
          </TimerProvider>
        </RightDynamicItemSetPairProvider>
      </LeftDynamicItemSetPairProvider>
    </div>
  );
};
