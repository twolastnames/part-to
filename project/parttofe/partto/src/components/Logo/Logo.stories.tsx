import React from "react";
import type { Meta, StoryObj } from "@storybook/react";

import { Logo, Size } from "./Logo";
import { ShellProvider } from "../../ShellProvider";

const meta: Meta<typeof Logo> = {
  component: Logo,
  decorators: (Story) => (
    <ShellProvider>
      <Story />
    </ShellProvider>
  ),
};
export default meta;

type Story = StoryObj<typeof Logo>;

export const Default: Story = {
  args: {},
};

export const Large: Story = {
  args: { size: Size.Large },
};
