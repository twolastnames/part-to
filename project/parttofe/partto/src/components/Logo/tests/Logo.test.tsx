import React from "react";
import { expect, test } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { ShellProvider } from "../../../ShellProvider";
import { Logo } from "../Logo";

test("snapshot", () => {
  render(
    <ShellProvider>
      <Logo />
    </ShellProvider>,
  );
  const component = screen.getByTestId("Logo");
  expect(component).toMatchSnapshot();
});
