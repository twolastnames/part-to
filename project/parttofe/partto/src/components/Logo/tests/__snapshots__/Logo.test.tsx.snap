// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="logo"
  data-testid="Logo"
  style="display: inline-block; height: 60px; overflow: hidden; width: 96px; font-family: \\"Courier New\\", Courier, monospace; font-size: 120px; line-height: 180px;"
>
  <div
    style="position: relative; top: -48px;"
  >
    <div
      style="height: 78px; overflow: hidden;"
    >
      <div>
        T
      </div>
    </div>
    <div
      style="height: 72px; overflow: hidden; position: relative; top: -48px; left: 24px;"
    >
      <div
        style="position: relative; top: -24px; width: 80.4px;"
      >
        o
      </div>
    </div>
  </div>
</div>
`;
